### PR TITLE
[Exchange] Refactor AbstractExchange

### PR DESCRIPTION
**Pull request by Herklos**
- Enable Future and Margin ExchangeSimulator
- Enable multiple different implementations of a rest exchange (for example, binance can now be implemented without ccxt)
- Enable potential non cryptocurrency exchange implementation
- Disable default ccxt implementation usage with margin / future exchange

Requires https://github.com/Drakkar-Software/OctoBot-Tentacles/pull/299

### CODE REVIEW

File: octobot_trading/exchanges/implementations/exchange_simulator.py [hunk 1]
```diff
@@ -128,7 +136,7 @@ def get_market_status(self, symbol, price_example=0, with_fixer=True):
     def get_uniform_timestamp(self, timestamp):
         return timestamp / 1000
 
-    def get_fees(self, symbol=None):
+    def get_fees(self, symbol):
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/implementations/exchange_simulator.py:139]: Is this not breaking tests ?

File: octobot_trading/exchanges/implementations/exchange_simulator.py [hunk 2]
```diff
@@ -156,8 +164,9 @@ def get_fees(self, symbol=None):
     #     'rate': percentage, // the fee rate, 0.05% = 0.0005, 1% = 0.01, ...
     #     'cost': feePaid, // the fee cost (amount * fee rate)
     # }
-    def get_trade_fee(self, symbol, order_type, quantity, price,
-                      taker_or_maker=ExchangeConstantsMarketPropertyColumns.TAKER.value):
+    def get_trade_fee(self, symbol, order_type, quantity, price, taker_or_maker):
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/implementations/exchange_simulator.py:167]: Is this not breaking tests ?

File: octobot_trading/exchanges/implementations/spot_ccxt_exchange.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+from ccxt import NotSupported, InsufficientFunds
+
+from octobot_trading import errors
+from octobot_trading.enums import TraderOrderType, ExchangeConstantsOrderColumns as ecoc
+from octobot_trading.errors import MissingFunds
+from octobot_trading.exchanges.implementations.ccxt_exchange import CCXTExchange
+from octobot_trading.exchanges.types.spot_exchange import SpotExchange
+
+
+class SpotCCXTExchange(CCXTExchange, SpotExchange):
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/implementations/spot_ccxt_exchange.py:26]: 👍

File: octobot_trading/producers/trades_updater.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
 #  License along with this library.
 import asyncio
 
-from ccxt.base.errors import NotSupported
+from octobot_trading.errors import NotSupported
```
Comment (GuillaumeDSM) [inline octobot_trading/producers/trades_updater.py:19]: 👍

**Review by GuillaumeDSM — COMMENTED** [review]
very nice improvement !

File: octobot_trading/exchanges/exchange_util.py [hunk 5]
```diff
@@ -56,4 +49,22 @@ def _search_exchange_class_from_exchange_type(exchange_type, exchange_class, ten
...
+    if enable_default:
+        return None
+
+    get_logger().warning(f"No specific exchange implementation for {exchange_name} found, searching for default one...")
+    return search_exchange_class_from_exchange_name(SpotExchange, exchange_name,
+                                                    tentacles_setup_config, enable_default=True)
+
+
+def _is_exchange_candidate_matching(exchange_candidate, exchange_name, tentacles_setup_config, enable_default=False):
+    return not exchange_candidate.is_simulated_exchange() and \
```
[the remark this answers is not in the document]
Comment (Herklos) [reply octobot_trading/exchanges/exchange_util.py:61]: No, I supposed we have only SpotExchangeSimulator, MarginExchangeSimulator and FutureExchangeSimulator possibilities for a backtesting exchange. If we want to add backtesting exchange tentacle feature we should adapt this method.

File: octobot_trading/exchanges/implementations/exchange_simulator.py (hunk 2 above)
Comment (Herklos) [reply octobot_trading/exchanges/implementations/exchange_simulator.py:167]: Exchange test are passing. I tested tentacles tests before this change, I'll have a look at tentacles tests

**Review by GuillaumeDSM — APPROVED** [review]
Good for me when we are sure tentacles tests are still fine !

**Pull request closed without merging**